nntrainer/dataset/batch_queue.cpp

```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   batch_queue.cpp
+ * @date   13 July 2021
+ * @brief  This file contains thread safe queue
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#include <batch_queue.h>
+#include <chrono>
+
+#include <nntrainer_error.h>
+
+using namespace std::literals::chrono_literals;
+
+namespace nntrainer {
+BatchQueue::BatchQueue(unsigned int queue_capacity_) :
+  queue_capacity(queue_capacity_) {
+  NNTR_THROW_IF(queue_capacity == 0, std::invalid_argument)
+    << "queue capacity of zero not supported!";
+}
+
+BatchQueue::BatchQueue(const BatchQueue &rhs) :
+  queue_capacity(rhs.queue_capacity) {}
+
+BatchQueue &BatchQueue::operator=(const BatchQueue &rhs) {
+  if (this == &rhs) {
+    return *this;
+  }
+  this->queue_capacity = rhs.queue_capacity;
+  return *this;
+}
+
+void BatchQueue::wait_and_push(T &&data) noexcept {
+  std::unique_lock<std::mutex> lk(q_mutex);
```

Comment:
as BatchQueue will be shared by the model and the dataset, I am guessing model will start dataset in a separate thread in this case?

Reply:
actually it is abstracted inside the databuffer, so model won't know about the existence of a queue.

please refer to the below code from #1416
https://github.com/nnstreamer/nntrainer/blob/71c7d99f6c533ae28e87513d6e81593d73ae4a8b/nntrainer/models/neuralnet.cpp#L687-L707

Reply:
Great, thanks for the explanation.